internal/v2/s2av2.go

```diff
@@ -0,0 +1,88 @@
+// Package s2av2 provides the S2Av2 transport credentials used by a gRPC
+// application.
+package s2av2
+
+import (
+	"errors"
+	"context"
+	"net"
+
+	"google.golang.org/grpc/credentials"
+	"github.com/google/s2a-go/internal/v2/tls_config_store"
+)
+
+const (
+	s2aSecurityProtocol = "s2av2"
+)
+
+type s2av2TransportCreds struct {
+	info     *credentials.ProtocolInfo
+	isClient bool
+}
+
+// NewClientCreds returns a client-side transport credentials object that uses
+// the S2Av2 to establish a secure connection with a server.
+func NewClientCreds() (credentials.TransportCredentials, error) {
+	creds := &s2av2TransportCreds{
+		info: &credentials.ProtocolInfo{
+			SecurityProtocol: s2aSecurityProtocol,
+		},
+		isClient: true,
+	}
+	return creds, nil
+}
+
+// NewServerCreds returns a server-side transport credentials object that uses
+// the S2Av2 to establish a secure connection with a client.
+func NewServerCreds() (credentials.TransportCredentials, error) {
+	creds := &s2av2TransportCreds{
+		info: &credentials.ProtocolInfo{
+			SecurityProtocol: s2aSecurityProtocol,
+		},
+		isClient: false,
+	}
+	return creds, nil
+}
+
+// ClientHandshake initiates a client-side mTLS handshake using the S2Av2.
+func (c *s2av2TransportCreds) ClientHandshake(ctx context.Context, serverAuthority string, rawConn net.Conn) (net.Conn, credentials.AuthInfo, error) {
+	if !c.isClient {
+		return nil, nil, errors.New("client handshake called using server transport credentials")
+	}
+	// TODO(rmehta19): Create a stub to S2Av2.
+	config := tlsconfigstore.GetTlsConfigurationForClient("serverHostname")
```

Comment:
Sorry I missed this before - we have to be a little careful with how we populate the server hostname parameter (which we currently have hardcoded to "serverHostname").

I think we need to do the following:

1. Have a new "serverAuthority" variable in the `s2av2TransportCredentials` struct.
2. In `ClientHandshake`, use the following logic: if `c.serverAuthority` is empty, pass the `serverAuthority` parameter to `GetTlsConfigurationForClient`; otherwise, pass in `c.serverAuthority`.
3. In `OverrideServerName`, also populate `c.serverAuthority` with the overridden server name.